### PR TITLE
PR: Include plugins in dependencies dialog

### LINKED ISSUE #11318: Add external plugins to dependencies
The idea is that this will allow the user to discover which plugins are installed. The list of plugins should also be included when using Spyder's issue reporter. Inspired by spyder-ide/spyder-line-profiler#39.

### PR DESCRIPTION
**Pull request by goanpeca**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

Screen Shot 2020-01-21 at 10 31 11

```

# Mandatory:
applaunchservices >=0.1.7     :  0.2.1 (OK)
atomicwrites >=1.2.0          :  1.3.0 (OK)
chardet >=2.0.0               :  3.0.4 (OK)
cloudpickle >=0.5.0           :  1.2.2 (OK)
diff_match_patch >=20181111   :  20181111 (OK)
intervaltree                  :  None (OK)
IPython >=4.0                 :  7.9.0 (OK)
jedi =0.14.1                  :  0.15.1 (NOK)
nbconvert >=4.0               :  5.6.1 (OK)
numpydoc >=0.6.0              :  0.9.1 (OK)
pexpect >=4.4.0               :  4.7.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=0.3                  :  5.6.7 (OK)
pygments >=2.0                :  2.5.2 (OK)
pylint >=0.25                 :  2.4.4 (OK)
pyls >=0.31.2;<0.32.0         :  0.31.1 (NOK)
qdarkstyle >=2.7              :  2.7 (OK)
qtawesome >=0.5.7             :  0.6.0 (OK)
qtconsole >=4.6.0             :  4.6.0 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.8.3                 :  0.8.3 (OK)
sphinx >=0.6.6                :  2.2.2 (OK)
spyder_kernels >=1.8.1;<2.0.0 :  1.9.0.dev0 (OK)
watchdog                      :  None (OK)
zmq >=17                      :  18.1.0 (OK)

# Optional:
cython >=0.21                 :  0.29.14 (OK)
matplotlib >=2.0.0            :  3.1.1 (OK)
numpy >=1.7                   :  1.17.4 (OK)
pandas >=0.13.1               :  0.25.3 (OK)
scipy >=0.17.0                :  1.3.1 (OK)
sympy >=0.7.3                 :  1.4 (OK)

# Spyder plugins:
spyder_notebook               :  0.3.0.dev0 (OK)
spyder_terminal               :  0.3.0.dev0 (OK)
```

### Issue(s) Resolved

Fixes #11318

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @goanpeca

### CODE REVIEW

**Comment by goanpeca** [comment]
Currently plugins do not provide some documentation, so we could either use the docstring of the module level or add a description parameter to the plugin?

File: spyder/dependencies.py [hunk 1]
```diff
@@ -290,9 +299,12 @@ def __init__(self, modname, package_name, features, required_version,
 
     def check(self):
         """Check if dependency is installed"""
-        return programs.is_module_installed(self.modname,
-                                            self.required_version,
-                                            self.installed_version)
+        if self.required_version and self.installed_version:
```
Comment (jitseniesen) [inline spyder/dependencies.py]: Why did you add this line?

File: spyder/widgets/dependencies.py [hunk 2]
```diff
@@ -29,32 +30,45 @@ def update_dependencies(self, dependencies):
...
-        optional_item = QTreeWidgetItem(["Optional"])
+
+        # Optional items
+        optional_item = QTreeWidgetItem([_("Optional")])
         optional_item.setFont(0, font)
-        self.addTopLevelItems([mandatory_item, optional_item])
+
+        # Spyder plugins
+        spyder_plugins = QTreeWidgetItem([_("Spyder Plugins")])
+
```
Comment (jitseniesen) [inline spyder/widgets/dependencies.py]: Add `spyder_plugins.setFont(0, font)` for consistency with other headers

File: spyder/dependencies.py (hunk 1 above)
Comment (goanpeca) [reply spyder/dependencies.py]: We do not have a "required version" for spyder plugins, so those deps do not define these values so I skip this check otherwise it would fails with a invalid version string comparison.

File: spyder/widgets/dependencies.py (hunk 2 above)
Comment (goanpeca) [reply spyder/widgets/dependencies.py]: Agree, thanks!

Comment (goanpeca) [reply spyder/widgets/dependencies.py]: You are correct! thanks

**Comment by jitseniesen** [comment]
> Currently plugins do not provide some documentation, so we could either use the docstring of the module level or add a description parameter to the plugin? For now I just used the plugin title

I would say, add a description parameter and use the plugin title as fallback. The description can be localized while the docstring is always English. Besides, the docstring of the module probably mentions somewhere that the module is to be used within Spyder, which will look strange in the dependencies dialog.

**Comment by goanpeca** [comment]
>Additionally, I noticed that in the dialog, the plugins appear at the bottom, but if you do "Copy to clipboard", they appear at the top. Not really important, but if it is easy to make this consistent then please do so.

Hmmm weird, will fix it :-) nice catch @jitseniesen !

File: spyder/widgets/dependencies.py (hunk 2 above)
Comment (goanpeca) [reply spyder/widgets/dependencies.py]: Fixed!

**Comment by goanpeca** [comment]
>I would say, add a description parameter and use the plugin title as fallback. The description can be localized while the docstring is always English. Besides, the docstring of the module probably mentions somewhere that the module is to be used within Spyder, which will look strange in the dependencies dialog.

@ccordoba12 do you agree with adding a `DESCRIPTION` or `get_description` to the plugins API, to provide information for this dialog?

**Comment by ccordoba12** [comment]
> do you agree with adding a DESCRIPTION or get_description to the plugins API, to provide information for this dialog?

Yes, I agree. And I think it's simpler to add DESCRIPTION to the `ATTRIBUTES` section of `SpyderPlugin`, instead of adding a method like `get_description`.

**Comment by goanpeca** [comment]
@ccordoba12 and @jitseniesen I updated the method that constructs the text so that dependencies are sorted by kind and name, which results in something like this:

```
# Mandatory:
applaunchservices >=0.1.7     :  0.2.1 (OK)
atomicwrites >=1.2.0          :  1.3.0 (OK)
chardet >=2.0.0               :  3.0.4 (OK)
cloudpickle >=0.5.0           :  1.2.2 (OK)
diff_match_patch >=20181111   :  20181111 (OK)
intervaltree                  :  None (OK)
IPython >=4.0                 :  7.9.0 (OK)
jedi =0.14.1                  :  0.15.1 (NOK)
nbconvert >=4.0               :  5.6.1 (OK)
numpydoc >=0.6.0              :  0.9.1 (OK)
pexpect >=4.4.0               :  4.7.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=0.3                  :  5.6.7 (OK)
pygments >=2.0                :  2.5.2 (OK)
pylint >=0.25                 :  2.4.4 (OK)
pyls >=0.31.2;<0.32.0         :  0.31.1 (NOK)
qdarkstyle >=2.7              :  2.7 (OK)
qtawesome >=0.5.7             :  0.6.0 (OK)
qtconsole >=4.6.0             :  4.6.0 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.8.3                 :  0.8.3 (OK)
sphinx >=0.6.6                :  2.2.2 (OK)
spyder_kernels >=1.8.1;<2.0.0 :  1.9.0.dev0 (OK)
watchdog                      :  None (OK)
zmq >=17                      :  18.1.0 (OK)
# Optional:
cython >=0.21                 :  0.29.14 (OK)
matplotlib >=2.0.0            :  3.1.1 (OK)
numpy >=1.7                   :  1.17.4 (OK)
pandas >=0.13.1               :  0.25.3 (OK)
scipy >=0.17.0                :  1.3.1 (OK)
sympy >=0.7.3                 :  1.4 (OK)
# Plugin:
spyder_notebook               :  0.3.0.dev0 (OK)
spyder_terminal               :  0.3.0.dev0 (OK)
```

Do you guys agree with this change?

**Comment by ccordoba12** [comment]
> Do you guys agree with this change?

Yes, looks nice! My only suggestion is to add a blank line between sections, i.e.}

```
# Mandatory:
applaunchservices >=0.1.7     :  0.2.1 (OK)
atomicwrites >=1.2.0          :  1.3.0 (OK)
chardet >=2.0.0               :  3.0.4 (OK)
cloudpickle >=0.5.0           :  1.2.2 (OK)
diff_match_patch >=20181111   :  20181111 (OK)
intervaltree                  :  None (OK)
IPython >=4.0                 :  7.9.0 (OK)
jedi =0.14.1                  :  0.15.1 (NOK)
nbconvert >=4.0               :  5.6.1 (OK)
numpydoc >=0.6.0              :  0.9.1 (OK)
pexpect >=4.4.0               :  4.7.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=0.3                  :  5.6.7 (OK)
pygments >=2.0                :  2.5.2 (OK)
pylint >=0.25                 :  2.4.4 (OK)
pyls >=0.31.2;<0.32.0         :  0.31.1 (NOK)
qdarkstyle >=2.7              :  2.7 (OK)
qtawesome >=0.5.7             :  0.6.0 (OK)
qtconsole >=4.6.0             :  4.6.0 (OK)
qtpy >=1.5.0                  :  1.9.0 (OK)
rtree >=0.8.3                 :  0.8.3 (OK)
sphinx >=0.6.6                :  2.2.2 (OK)
spyder_kernels >=1.8.1;<2.0.0 :  1.9.0.dev0 (OK)
watchdog                      :  None (OK)
zmq >=17                      :  18.1.0 (OK)

# Optional:
cython >=0.21                 :  0.29.14 (OK)
matplotlib >=2.0.0            :  3.1.1 (OK)
numpy >=1.7                   :  1.17.4 (OK)
pandas >=0.13.1               :  0.25.3 (OK)
scipy >=0.17.0                :  1.3.1 (OK)
sympy >=0.7.3                 :  1.4 (OK)

# Plugin:
spyder_notebook               :  0.3.0.dev0 (OK)
spyder_terminal               :  0.3.0.dev0 (OK)
```

**Comment by ccordoba12** [comment]
Well, and change the header of plugins section to be

```
# Spyder plugins:
spyder_notebook               :  0.3.0.dev0 (OK)
spyder_terminal               :  0.3.0.dev0 (OK)
```

**Comment by goanpeca** [comment]
@ccordoba12 @jitseniesen the [description of the issue](https://github.com/spyder-ide/spyder/pull/11364#issue-365083929) has the final look

**Comment by ccordoba12** [comment]
This looks good for me, thanks @goanpeca!

@jitseniesen?
